File: examples/02-reliability/python/partial_failure.py
import asyncio
from datetime import timedelta

from temporalio import workflow
from temporalio.exceptions import ActivityError

# `process_with_status` is a module-level @activity.defn function.


@workflow.defn
class PartialFailureWorkflow:
    # gather(return_exceptions=True) lets some branches fail without aborting the
    # whole fan-out: failures arrive as exception objects in the result list, so
    # the Workflow can record a per-partition status instead of crashing.
    @workflow.run
    async def process(self, partitions: list[int]) -> dict[int, str]:
        futures = {
            partition: workflow.execute_activity(
                process_with_status,
                partition,
                start_to_close_timeout=timedelta(minutes=15),
            )
            for partition in partitions
        }
        outcomes = await asyncio.gather(*futures.values(), return_exceptions=True)

        result: dict[int, str] = {}
        for partition, outcome in zip(futures.keys(), outcomes):
            if isinstance(outcome, ActivityError):
                result[partition] = f"FAILED: {outcome}"
            else:
                result[partition] = outcome
        return result

File: examples/02-reliability/go/partial_failure.go
package reliability

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/workflow"
)

// Let some partitions fail without aborting the whole fan-out. We start every
// Activity, then Get each Future and record a per-partition status — a failed
// Get returns an error here instead of unwinding the Workflow.
func Process(ctx workflow.Context, partitions []int) (map[int]string, error) {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 15 * time.Minute,
	})

	futures := make(map[int]workflow.Future, len(partitions))
	for _, partition := range partitions {
		futures[partition] = workflow.ExecuteActivity(ctx, "ProcessWithStatus", partition)
	}

	result := make(map[int]string, len(partitions))
	for partition, f := range futures {
		var status string
		if err := f.Get(ctx, &status); err != nil {
			result[partition] = fmt.Sprintf("FAILED: %v", err)
		} else {
			result[partition] = status
		}
	}
	return result, nil
}
